In [1]:
!pip install transformers torch torchvision torchaudio tensorflow 

In [2]:
from transformers import pipeline, set_seed, GPT2Tokenizer

In [3]:
import os
import nltk

In [4]:
file_path = "unit 1.txt"

In [5]:
try:
    with open(file_path, "r", encoding="utf-8") as f:
        text = f.read()
    print("File loaded successfully!")
except FileNotFoundError:
    print(f"Error: '{file_path}' not found.")

File loaded successfully!


# Experiment 1: Text Generation

In [7]:
set_seed(45)

In [6]:
prompt = "The future of AI is"

### BERT

In [10]:
fast_generator = pipeline('text-generation', model="bert-base-uncased")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`
Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=25) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


The future of AI is................................................................................................................................................................................................................................................................


### RoBERTa

In [11]:
fast_generator = pipeline('text-generation', model="roberta-base")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

C:\Users\Ritu\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Ritu\.cache\huggingface\hub\models--roberta-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=25) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


The future of AI is


### BART

In [13]:
fast_generator = pipeline('text-generation', model="facebook/bart-base")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

Some weights of BartForCausalLM were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=25) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


The future of AI is filedв Verge571 Trip Trip Tripcodes filed thruPassword LearnfcElsElsEls thru thru tumult thru generously Brass thruPassword FactIFT NIGHTounty529ElsEls SUM thru thru method FactEls thru itching itchingElsEls arrested needles bugsIFTIFTPasswordughIFT thru thru Jim bugs170 ScaleIFT NIGHTogenicogenic NIGHT biIFT dilemmaodies Beaver Muss NIGHT NIGHT Verb endors twist twist Scale VerbPassword clusters Fact Scaleodies biStrong thru endors ScaleIFTIFT piercingStrongStrongPassword itching lineup bi signaling establishmentogenic NIGHTogenicatherStrong frontline lineupogenicogenic lineup jarStrongStrongEnergy SyndicateStrong thriller battledogenic bi PlumStrongStrong testim testimStrong biogenicogenic biStrongogenicodiesogenic inept bi FortyStrongogenicogenicogenicounty sane bi bi failingogenic perpetratedogenic NIGHTStrongStrongresponsible piercing Fact twist invalidStrong SyndicateStrongEls NIGHT testimogeniccodesogenicStrong thrillerogenicogenic IDF patriarchalStrongodies 

# Experiment 2: Masked Language Modeling (Missing Word)

### BERT

In [16]:
mask_filler = pipeline("fill-mask", model="bert-base-uncased")

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


In [17]:
masked_sentence = "The goal of Generative AI is to [MASK] new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

create: 0.54
generate: 0.16
produce: 0.05
develop: 0.04
add: 0.02


### RoBERTa

In [18]:
mask_filler = pipeline("fill-mask", model="roberta-base")

Device set to use cpu


In [21]:
masked_sentence = "The goal of Generative AI is to <mask> new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

 generate: 0.37
 create: 0.37
 discover: 0.08
 find: 0.02
 provide: 0.02


### BART

In [23]:
mask_filler = pipeline("fill-mask", model="facebook/bart-base")  

Device set to use cpu


In [24]:
masked_sentence = "The goal of Generative AI is to <mask> new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

 create: 0.07
 help: 0.07
 provide: 0.06
 enable: 0.04
 improve: 0.03


# Experiment 3: Question Answering

In [27]:
question = "What are the risks?"
context = "Generative AI poses significant risks such as hallucinations, bias, and deepfakes."

### BERT

In [28]:
qa_pipeline = pipeline("question-answering", model="bert-base-uncased")

Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


In [29]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: as hallucinations


### RoBERTa

In [30]:
qa_pipeline = pipeline("question-answering", model="roberta-base")

Some weights of RobertaForQuestionAnswering were not initialized from the model checkpoint at roberta-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


In [31]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: Generative AI poses significant risks such as hallucinations, bias, and deepfakes


### BART

In [32]:
qa_pipeline = pipeline("question-answering", model="facebook/bart-base")

Some weights of BartForQuestionAnswering were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


In [33]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: Generative AI


# Observation Table

| Task | Model | Classification (Success/Failure) | Observation (What actually happened?) | Why did this happen? (Architectural Reason) |
| :--- | :--- | :--- | :--- | :--- |
| **Generation** | BERT | *Failure* | *Generated dots instead of meaningful text.* | *BERT is an Encoder; it isn't trained to predict the next word.* |
| | RoBERTa | *Failure* | *Returned only the prompt without continuation.* | *RoBERTa is also encoder-only and cannot generate new tokens.* |
| | BART | *Failure* | *Produced long, incoherent, random-looking text.* | *BART is encoder–decoder, but the text-generation pipeline expects a decoder-only (causal LM), so it was misused.* |
| **Fill-Mask** | BERT | *Success* | *Predicted create (0.54), generate (0.16), produce (0.05).* | *BERT is trained on Masked Language Modeling (MLM).* |
| | RoBERTa | *Success* | *Predicted generate (0.37), create (0.37), discover (0.08).* | *RoBERTa is also trained with MLM, optimized over BERT.* |
| | BART | *Partial Success* | *Predicted weaker and more generic words like create (0.07), help (0.07), provide (0.06)* | *BART is trained with denoising autoencoding, but not optimized for single-token MLM.* |
| **QA** | BERT | *Partial Success* | *Returned fragment: “as hallucinations”* | *Base BERT is not fine-tuned for extractive QA, so span prediction is weak.* |
| | RoBERTa | *Success* | *Returned full correct answer: hallucinations, bias, and deepfakes* | *Even without fine-tuning, RoBERTa’s representations are strong enough for short contexts.* |
| | BART | *Failure* | *Returned incomplete answer: “Generative AI”* | *BART-base is not trained for extractive QA and struggles with span extraction.* |